namesys/dns.go

```diff
@@ -41,33 +41,66 @@ func (r *DNSResolver) ResolveN(ctx context.Context, name string, depth int) (pat
 	return resolve(ctx, r, name, depth, "/ipns/")
 }
 
+type lookupRes struct {
+	path path.Path
+	error error
+}
+
 // resolveOnce implements resolver.
 // TXT records for a given domain name should contain a b58
 // encoded multihash.
 func (r *DNSResolver) resolveOnce(ctx context.Context, name string) (path.Path, error) {
 	segments := strings.SplitN(name, "/", 2)
+	domain := segments[0]
 
-	if !isd.IsDomain(segments[0]) {
+	if !isd.IsDomain(domain) {
 		return "", errors.New("not a valid domain name")
 	}
+	log.Infof("DNSResolver resolving %s", domain)
+
+	rootChan := make(chan lookupRes, 1)
+	go workDomain(r, domain, rootChan)
+
+	subChan := make(chan lookupRes, 1)
+	go workDomain(r, "_dnslink." + domain, subChan)
+
+	subRes := <- subChan
```

Comment:
sorry to hang this PR up any longer, but if you could change your channel receives to also check the context, that would be super fantastic:
```go
var subRes *lookupRes
select {
    case subRes = <-subChan:
    case <-ctx.Done():
        return "", ctx.Err()
}
```

do that for this one and the rootChan one a little lower and i'll merge asap